core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala

```diff
@@ -729,7 +729,7 @@ private[spark] class ExternalSorter[K, V, C](
       var out: FileOutputStream = null
       var in: FileInputStream = null
       val writeStartTime = System.nanoTime
-      try {
+      util.Utils.tryWithSafeFinally {
```

Comment:
I guess the same comment goes here, although I see it's kind of because it's trying to clean up two streams but skimping on finally blocks, with just one to do both.

But there's really no reason that the input stream has to be outside the for loop. It should be opened and closed locally inside the loop over partitions. So I think this can be usefully simplified along the same lines -- `out` declared outside the try block, and push `in` inside that loop.